example_scripts/download_vhi_parallel.py

```diff
@@ -0,0 +1,162 @@
+        ftp.cwd('/pub/corp/scsb/wguo/data/Blended_VH_4km/VH/')
+
+        # append the filenames to a list
+        listing = []
+        ftp.retrlines("LIST", listing.append)
+        # extract the filename
+        filepaths = [f.split(' ')[-1] for f in listing]
+        # extract only the filenames of interest
+        vhi_files = [f for f in filepaths if ".VH.nc" in f]
+
+    return vhi_files
+
+
+def download_file_from_ftp(ftp_instance, filename, output_filename):
+    """ by having context manager INSIDE function
+    TODO: is this sensible? or have the FTP as an argument to be supplied ?
+    """
+    print(f"Downloading {output_filename}")
+    with open(output_filename,'wb') as lf:
+        ftp_instance.retrbinary("RETR " + filename, lf.write)
+    if output_filename.exists():
+        print(f"Successful Download! {output_filename}")
+    else:
+        print(f"Error Downloading file: {output_filename}")
+
+    return
+
+
+def batch_ftp_request(filenames, output_dir=OUTPUT_DIR):
+    """ This has the context manager to avoid overloading ftp server """
+    with FTP('ftp.star.nesdis.noaa.gov') as ftp:
+        ftp.login()
+        ftp.cwd('/pub/corp/scsb/wguo/data/Blended_VH_4km/VH/')
+
+        for filename in filenames:
+            output_filename = output_dir / filename
+            download_file_from_ftp(ftp, filename, output_filename)
+
+    return
+
+
+def each_file_individually(filenames, output_dir=OUTPUT_DIR):
+    """ individual context for each file (will that overload that poor server?)"""
+    assert False, "Is this irresponsible?"
+    with FTP('ftp.star.nesdis.noaa.gov') as ftp:
+        ftp.login()
+        ftp.cwd('/pub/corp/scsb/wguo/data/Blended_VH_4km/VH/')
+        for filename in filenames:
+            output_filename = output_dir / filename
+            download_file_from_ftp(ftp, filename, output_filename)
+    return
+
+
+
+def chunks(l, n):
+    """ return a generator object which chunks list into sublists of size n
+    https://chrisalbon.com/python/data_wrangling/break_list_into_chunks_of_equal_size/
+    """
+    # For item i in a range that is a length of l,
+    for i in range(0, len(l), n):
+        # Create an index range for l of n items:
+        yield l[i:i+n]
+
+
+def main(pool, subset=False):
+    # get the filenames
+    if subset:
+        vhi_files = get_ftp_filenames()[:15]
+    else:
+        vhi_files = get_ftp_filenames()
+
+    # split the filenames into batches of 100 (21 batches)?
+    batches = [batch for batch in chunks(vhi_files,100)]
+
+    # run in parallel for multiple file downloads
+    # pool = multiprocessing.Pool(processes=100)
+    ris = pool.map(batch_ftp_request, batches)
+
+    # write the output (TODO: turn into logging behaviour)
+    print("\n\n*************************\n\n")
+    print("Script Run")
+    print("*************************")
+    print("Errors:")
+    print("\nError: ",[ri for ri in ris if ri != None])
+
+
+def test(parallel=True, pool=None):
```

Comment:
@gabrieltseng keep or delete? not maintainable functionality imo but maybe useful?